cmd/ecr-credential-provider/main.go

```diff
@@ -0,0 +1,163 @@
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package main
+
+import (
+	"context"
+	"encoding/base64"
+	"errors"
+	"fmt"
+	"net/url"
+	"os"
+	"regexp"
+	"strings"
+	"time"
+
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/aws/session"
+	"github.com/aws/aws-sdk-go/service/ecr"
+
+	plugin "k8s.io/cloud-provider-aws/cmd/ecr-credential-provider/framework"
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/klog/v2"
+	"k8s.io/kubelet/pkg/apis/credentialprovider/v1alpha1"
+)
+
+var ecrPattern = regexp.MustCompile(`^(\d{12})\.dkr\.ecr(\-fips)?\.([a-zA-Z0-9][a-zA-Z0-9-_]*)\.(amazonaws\.com(\.cn)?|sc2s\.sgov\.gov|c2s\.ic\.gov)$`)
+
+type ECR interface {
+	GetAuthorizationToken(input *ecr.GetAuthorizationTokenInput) (*ecr.GetAuthorizationTokenOutput, error)
+}
+
+type ECRProvider interface {
+	NewECR(region string, registryID string) (*ecr.ECR, error)
+}
+
+type ecrPlugin struct {
+	ecrProvider ECRProvider
+	ecr         ECR
+}
+
+func defaultECRProvider(region string, registryID string) (*ecr.ECR, error) {
+	sess, err := session.NewSessionWithOptions(session.Options{
+		Config:            aws.Config{Region: aws.String(region)},
+		SharedConfigState: session.SharedConfigEnable,
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	return ecr.New(sess), nil
+}
+
+func (e *ecrPlugin) GetCredentials(ctx context.Context, image string, args []string) (*v1alpha1.CredentialProviderResponse, error) {
+	registryID, region, registry, err := parseRepoURL(image)
+	if err != nil {
+		return nil, err
+	}
+
+	if e.ecr == nil {
+		e.ecr, err = defaultECRProvider(region, registryID)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	output, err := e.ecr.GetAuthorizationToken(&ecr.GetAuthorizationTokenInput{
+		RegistryIds: []*string{aws.String(registryID)},
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	if output == nil {
+		return nil, errors.New("response output from ECR was nil")
+	}
+
+	if len(output.AuthorizationData) == 0 {
+		return nil, errors.New("authorization data was nil")
+	}
+
+	data := output.AuthorizationData[0]
+	if data.AuthorizationToken == nil {
+		return nil, errors.New("authorization token in response was nil")
+	}
+
+	decodedToken, err := base64.StdEncoding.DecodeString(aws.StringValue(data.AuthorizationToken))
+	if err != nil {
+		return nil, err
+	}
+
+	parts := strings.SplitN(string(decodedToken), ":", 2)
+	if len(parts) != 2 {
+		return nil, errors.New("error parsing username and password from authorization token")
+	}
+
+	var cacheDuration *metav1.Duration
+	expiresAt := data.ExpiresAt
+	if expiresAt == nil {
+		// explicitly set cache duration to 0 if expiresAt was nil so that
+		// kubelet does not cache it in-memory
+		cacheDuration = &metav1.Duration{Duration: 0}
+	} else {
+		duration := expiresAt.Sub(time.Now())
```

Comment:
Good callout. Actually since token comes from an makes an API call, we need to consider that latency as well, which would more random. Is it too late to change the `CredentialProviderResponse` to pass the `CacheExpiresAt` directly instead of `CacheDuration`?

Reply:
For compatibility, we can add it as an optional parameter and it'd take precedence over `CacheDuration`.

Reply:
I've substracted 1 second for now, which I think should be good for most cases. We can increase it if we want to be safer.

Reply:
Even 1 second seems pretty aggressive, can we subtract maybe an hour? That gives plenty of headroom in case there are errors retrieving new tokens but not too short of a duration where we're unnecessarily fetching tokens. 

Even half the expiration from ECR seems reasonable to me but I'll defer to @nckturner on what is appropriate.